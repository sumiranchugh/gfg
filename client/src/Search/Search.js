import React from "react";
import { Form, InputGroup } from "react-bootstrap";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faSearch } from "@fortawesome/free-solid-svg-icons";
import { filterProductsAPI, fetchProductsAPI } from "../common/apis/productApi";

class Search extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      changed: false
    };
  }
  handleSearch = e => {
    const searchTxt = e.target.value;
    if (searchTxt === "" && this.state.changed) {
      this.setState({ changed: false });
      fetchProductsAPI(this.props.updateProducts, this.props.errorCallback);
    } else if (searchTxt.length > 3) {
      this.setState({ changed: true });
      filterProductsAPI(
        searchTxt,
        this.props.updateProducts,
        this.props.errorCallback
      );
    }
  };
  render() {
    return (
      <Form>
        <Form.Group>
          <InputGroup>
            <Form.Control
              size="lg"
              type="text"
              placeholder="Search Our Catalogue"
              className="border-dark"
              onChange={this.handleSearch}
              aria-describedby="btnGroupAddon"
            ></Form.Control>
            <InputGroup.Append>
              <InputGroup.Text id="btnGroupAddon">
                <FontAwesomeIcon icon={faSearch} />
              </InputGroup.Text>
            </InputGroup.Append>
          </InputGroup>
        </Form.Group>
      </Form>
    );
  }
}

export default Search;
