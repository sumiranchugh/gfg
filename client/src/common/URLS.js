export const PRODUCTS = "api/products/";
export const PRODUCTS_FILTER = "api/products/filter?q=";
