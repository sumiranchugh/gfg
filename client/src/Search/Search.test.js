import React from "react";
import Adapter from "enzyme-adapter-react-16";
import renderer from "react-test-renderer";
import Search from "./Search";
import expect from "expect";
import { Form } from "react-bootstrap";
import { shallow, mount, render, configure } from "enzyme";
import { PRODUCTS_FILTER, PRODUCTS } from "../common/URLS";
configure({ adapter: new Adapter() });

// it("renders without crashing", () => {
//   const tree = renderer.create(<App/>).toJSON();
//   expect(tree).toMatchSnapshot();
// });

describe("Search", () => {
  it("render search with no text", done => {
    const wrapper = shallow(<Search />);
    expect(wrapper).toMatchSnapshot();
    process.nextTick(() => {
      expect(wrapper.state().changed).toEqual(false);
    });
    done();
  });
  it("render search with text length greater than 3", done => {
    const mockSuccessResponse = [];
    const mockJsonPromise = Promise.resolve(mockSuccessResponse);
    const mockFetchPromise = Promise.resolve({
      json: () => mockJsonPromise
    });
    global.fetch = jest.fn().mockImplementation(() => mockFetchPromise);

    const event = {
      preventDefault() {},
      target: { value: "the-value" }
    };
    const wrapper = mount(<Search />);
    wrapper.find(Form.Control).simulate("change", event);
    expect(global.fetch).toHaveBeenCalledTimes(1);
    expect(global.fetch).toHaveBeenCalledWith(
      "api/products/filter?q=the-value"
    );
    process.nextTick(() => {
      expect(wrapper.state().changed).toEqual(true);
      // expect(wrapper.find('Products').props().products).toEqual(mockSuccessResponse);
    });

    expect(wrapper).toMatchSnapshot();
    global.fetch.mockClear();
    delete global.fetch;

    done();
  });
  it("render app with backend error", done => {
    global.fetch = jest.fn().mockRejectedValue("");
    const mock = jest.fn();
    const event = {
      preventDefault() {},
      target: { value: "the-value" }
    };
    const wrapper = shallow(<Search />);
    wrapper.setProps({ errorCallback: mock });
    wrapper.find(Form.Control).simulate("change", event);
    expect(global.fetch).toHaveBeenCalledTimes(1);
    expect(global.fetch).toHaveBeenCalledWith(
      "api/products/filter?q=the-value"
    );
    process.nextTick(() => {
      expect(wrapper.state().changed).toEqual(true);
      expect(mock).toHaveBeenCalledTimes(1);
    });
    expect(wrapper).toMatchSnapshot();
    global.fetch.mockClear();
    delete global.fetch;
    done();
  });
});

it("render search with text length less than equal 3", done => {
  global.fetch = jest.fn();

  const wrapper = mount(<Search />);
  wrapper.find(Form.Control).simulate("change", {});
  expect(global.fetch).toHaveBeenCalledTimes(0);
  process.nextTick(() => {
    expect(wrapper.state().changed).toEqual(false);
  });

  expect(wrapper).toMatchSnapshot();
  global.fetch.mockClear();
  delete global.fetch;

  done();
});

it("render search with text length from more than  3 to 0", done => {
  const mockSuccessResponse = [];
  const mockJsonPromise = Promise.resolve(mockSuccessResponse);
  const mockFetchPromise = Promise.resolve({
    json: () => mockJsonPromise
  });
  global.fetch = jest.fn().mockImplementation(() => mockFetchPromise);

  const event = {
    preventDefault() {},
    target: { value: "" }
  };
  const wrapper = mount(<Search />);
  wrapper.setState({ changed: true });
  wrapper.find(Form.Control).simulate("change", event);
  expect(global.fetch).toHaveBeenCalledTimes(1);
  expect(global.fetch).toHaveBeenCalledWith(PRODUCTS);
  process.nextTick(() => {
    expect(wrapper.state().changed).toEqual(false);
  });

  expect(wrapper).toMatchSnapshot();
  global.fetch.mockClear();
  delete global.fetch;

  done();
});
