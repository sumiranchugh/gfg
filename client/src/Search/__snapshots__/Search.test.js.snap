// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search render app with backend error 1`] = `ShallowWrapper {}`;

exports[`Search render search with no text 1`] = `ShallowWrapper {}`;

exports[`Search render search with text length greater than 3 1`] = `ReactWrapper {}`;

exports[`render search with text length from more than  3 to 0 1`] = `ReactWrapper {}`;

exports[`render search with text length less than equal 3 1`] = `ReactWrapper {}`;
